homeassistant/components/tag/__init__.py

```diff
@@ -0,0 +1,136 @@
+"""The Tag integration."""
+import copy
+import datetime as dt
+import logging
+import typing
+from typing import Optional, cast
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_ID, CONF_NAME
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import collection
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.storage import Store
+from homeassistant.loader import bind_hass
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+DEVICE_ID = "device_id"
+EVENT_TAG_SCANNED = "tag_scanned"
+LAST_SCANNED = "last_scanned"
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+TAG_ID = "tag_id"
+TAGS = "tags"
+
+CREATE_FIELDS = {
+    vol.Required(CONF_ID): cv.string,
+    vol.Optional(CONF_NAME): vol.All(str, vol.Length(min=1)),
+    vol.Optional("description"): cv.string,
+    vol.Optional(LAST_SCANNED): cv.datetime,
+}
+
+UPDATE_FIELDS = {
+    vol.Optional(CONF_NAME): vol.All(str, vol.Length(min=1)),
+    vol.Optional("description"): cv.string,
+    vol.Optional(LAST_SCANNED): cv.datetime,
+}
+
+
+class TagIDExistsError(HomeAssistantError):
+    """Raised when an item is not found."""
+
+    def __init__(self, item_id: str):
+        """Initialize tag id exists error."""
+        super().__init__(f"Tag with id: {item_id} already exists.")
+        self.item_id = item_id
+
+
+class TagIDManager(collection.IDManager):
+    """ID manager for tags."""
+
+    def generate_id(self, suggestion: str) -> str:
+        """Generate an ID."""
+        if self.has_id(suggestion):
+            raise TagIDExistsError(suggestion)
+
+        return suggestion
+
+
+class TagStorageCollection(collection.StorageCollection):
+    """Tag collection stored in storage."""
+
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+    UPDATE_SCHEMA = vol.Schema(UPDATE_FIELDS)
+
+    async def _process_create_data(self, data: typing.Dict) -> typing.Dict:
+        """Validate the config is valid."""
+        if TAG_ID in data:
+            data[CONF_ID] = data.pop(TAG_ID)
+        return self.CREATE_SCHEMA(data)
+
+    @callback
+    def _get_suggested_id(self, info: typing.Dict) -> str:
+        """Suggest an ID based on the config."""
+        return info[CONF_ID]
+
+    async def _update_data(self, data: dict, update_data: typing.Dict) -> typing.Dict:
+        """Return a new updated data object."""
+        update_data = self.UPDATE_SCHEMA(update_data)
+        return {**data, **update_data}
+
+    async def _async_load_data(self) -> Optional[dict]:
+        """Load the data."""
+        items = cast(Optional[dict], await self.store.async_load())
+        if items:
+            for item in items.values():
+                if "last_scanned" in item:
```

Comment:
No need to check this, it should always be there.

Reply:
I was thinking that we may have a helper like UI where folks could manually register tags. If that is the case this maybe needs to be optional.

Reply:
dropped in favor of solution given below